### PR TITLE
Harden the freezing procedure against failures

### PR DESCRIPTION
**Pull request by adaszko**
Follow up to https://github.com/sigp/lighthouse/pull/1285#discussion_r445471441

Should be the last PR related to atomicity (https://github.com/sigp/lighthouse/issues/692).

### CODE REVIEW

File: beacon_node/store/src/hot_cold_store.rs [hunk 1]
```diff
@@ -792,45 +778,95 @@ pub fn process_finalization<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>>(
...
+    // Since it is pretty much impossible to be atomic across more than one database, we trade
+    // losing track of states to delete, for consistency.  In other words: We should be safe to die
+    // at any point below but it may happen that some states won't be deleted from the hot database
+    // and will remain there forever.  Since dying in these particular few lines should be an
+    // exceedingly rare event, this should be an acceptable tradeoff.
+
+    // Flush to disk all the states that have just been migrated to the cold store.
+    store
+        .cold_db
+        .put_bytes_fsync("fsync", "fsync".as_bytes(), "fsync".as_bytes())?;
```
Comment (michaelsproul) [inline beacon_node/store/src/hot_cold_store.rs:828]: This looks quite weird, could we wrap it in a method? Something like `store.cold_db.fsync()`?

File: beacon_node/beacon_chain/tests/store_tests.rs [hunk 2]
```diff
@@ -40,7 +44,11 @@ fn get_store(db_path: &TempDir) -> Arc<HotColdDB<E, LevelDB<E>, LevelDB<E>>> {
     let hot_path = db_path.path().join("hot_db");
     let cold_path = db_path.path().join("cold_db");
     let config = StoreConfig::default();
-    let log = NullLoggerBuilder.build().expect("logger should build");
+
+    let decorator = slog_term::PlainDecorator::new(slog_term::TestStdoutWriter);
+    let drain = slog_term::FullFormat::new(decorator).build();
+    let log = slog::Logger::root(std::sync::Mutex::new(drain).fuse(), o!());
```
[the remark this answers is not in the document]
Comment (adaszko) [reply beacon_node/beacon_chain/tests/store_tests.rs:50]: Yep, it's meant to be used with `--nocapture`.

[the remark this answers is not in the document]
Comment (michaelsproul) [reply beacon_node/beacon_chain/tests/store_tests.rs:50]: Sounds good!

File: beacon_node/store/src/hot_cold_store.rs (hunk 1 above)
Comment (adaszko) [reply beacon_node/store/src/hot_cold_store.rs]: Done!

File: beacon_node/store/src/hot_cold_store.rs [hunk 3]
```diff
@@ -792,45 +778,95 @@ pub fn process_finalization<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>>(
...
 
         // Store a pointer from this state root to its slot, so we can later reconstruct states
         // from their state root alone.
-        store.store_cold_state_slot(&state_root, slot)?;
+        let cold_state_summary = ColdStateSummary { slot };
+        let op = cold_state_summary.as_kv_store_op(state_root);
+        cold_db_ops.push(op);
+
+        // There are data dependencies between calls to `store_cold_state()` that prevent us from
+        // doing one big call to `store.cold_db.do_atomically()` at end of the loop.
```
[the remark this answers is not in the document]
Comment (adaszko) [reply beacon_node/store/src/hot_cold_store.rs:808]: That's right.   In fact I don't see any gain at all!  Even if we put everything in a single `do_atomically()` call, we'd still be exposed to the problem of keeping some states in the hot db indefinitely.  Moreover, if the OS process dies between some calls to `store.cold_db.do_atomically()`, the freezing process will simply be retried automatically and the states that have been written so far will simply be overwritten, so all fine and dandy.

**Pull request merged**